rust/parutils/src/buffered_async_iterator.rs

```diff
@@ -0,0 +1,395 @@
+                    data_queue.push(BufferItem::Completed).await;
+                    completion_flag.store(true, std::sync::atomic::Ordering::SeqCst);
+                    break;
+                };
+                data_queue.push(BufferItem::Value(v)).await;
+            }
+
+            if let Some(sendback) = stream_sendback.take() {
+                if sendback.send(src_iter).is_err() {
+                    info!("Error sending iterator back on tokio stream; receiver closed.");
+                }
+            }
+        })
+    }
+
+    /// Forces the background task to run to completion.  Used in testing. to ensure that the
+    /// completion flag has been properly set.  Should not be needed in most cases.
+    pub async fn cleanup_background_task(&mut self) -> Result<(), JoinError> {
+        if let Some(jh) = self.background_handle.take() {
+            jh.await?;
+        }
+        Ok(())
+    }
+}
+
+impl<It: AsyncIterator<E>, E: Send + Sync + 'static> Drop for BufferedAsyncIterator<It, E> {
+    fn drop(&mut self) {
+        if let Some(jh) = self.background_handle.take() {
+            jh.abort();
+        }
+    }
+}
+
+#[async_trait]
+impl<It: AsyncIterator<E>, E: Send + Sync + 'static> AsyncIterator<E>
+    for BufferedAsyncIterator<It, E>
+{
+    type Item = It::Item;
+
+    /// The traditional next method; returns None if everything is done.
+    async fn next(&mut self) -> Result<Option<Self::Item>, E> {
+        match self.data_queue.pop().await {
+            BufferItem::Value(v) => Ok(Some(v)),
+            BufferItem::Completed => Ok(None),
+            BufferItem::Error(e) => Err(e),
+        }
+    }
+}
+
+#[async_trait]
+impl<It: AsyncIterator<E>, E: Send + Sync + 'static> BatchedAsyncIterator<E>
+    for BufferedAsyncIterator<It, E>
+{
+    async fn next_batch(&mut self, max_num: Option<usize>) -> Result<Vec<Self::Item>, E> {
+        let n_ready = self.data_queue.len();
+
+        // If there are None ready currently, do we wait for more or be done with it?
+        if n_ready == 0 && self.all_items_in_buffer() {
+            return Ok(Vec::new());
+        }
+
+        // Now, there will be at least one coming down the line, so block on that.
+        let n_ready = n_ready.max(1);
+
+        // We have a batch ready, so pull them all in.
+        let n_retrieve = max_num.unwrap_or(n_ready).min(n_ready);
+        let mut ret = Vec::with_capacity(n_retrieve);
+
+        for _ in 0..n_retrieve {
+            match self.data_queue.pop().await {
+                BufferItem::Value(v) => ret.push(v),
+                BufferItem::Completed => break,
+                BufferItem::Error(e) => return Err(e),
+            };
+        }
+        Ok(ret)
+    }
+
+    /// If known, returns the number of items remaining.
+    fn items_remaining(&self) -> Option<usize> {
+        if self.all_items_in_buffer() {
+            let n = self.data_queue.len();
+            Some(n.saturating_sub(1))
```

Comment:
A bit confused why subtracting 1 here?

Reply:
Ah, it's because there's a `Complete` at the end. Let's put a comment here to explain.